### ISSUE TITLE
ampq:connection:forced disconnect due to idle. Can't set idle time. Regression bug.

### ISSUE THREAD

**Issue by icfly2** [Azure/azure-sdk-for-python#34212]
- **Package Name**: 
- azure-servicebus
- **Package Version**: 
- 7.11.4
- **Operating System**:
- linux (but bug is independent of that)
- **Python Version**: 
- 3.10 adn 3.11

**Describe the bug**
After #30269 the `idle_timeout` of the service bus sdk cannot be set but forces a disconnect after some time.

**To Reproduce**
1. upgrade from 7.9.0 to any version after #30296 such as latest 7.11.4.

**Expected behavior**
The connection should stay connected if no idle timeout is set. As per previous behaviour.

**Screenshots**

**Additional context**
This was already raised in #9917 then fixed at some point and now regressed again.

Object is initialised in:  sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_client.py:104-152 
But you can't set the corresponding kwargs

The disconnect is triggered in  sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp_transport.py:45-47 the code prior to the last edit of that function was working fine. Thus the pinpointing to the above PR.

**Comment by kashifkhan** [comment]
Thank you for the feedback @icfly2 . Can you provide some sample code to reproduce ? For example are you using amqp transport or websockets ? Are there any logs that can be provided as well that can help us debug this issue.

**Comment by icfly2** [comment]
This is the log message:Connection closed with error: ["amqp:connection:forced","The connection was closed by container '44d7f1a82b314fcda6c15e9c24c45d1d_G5' because it did not have any active links in the past 300000 milliseconds. TrackingId:44d7f1a82b314fcda6c15e9c24c45d1d_G5, SystemTracker:gateway5, Timestamp:2024-02-03T17:59:17",null]The code running is the azure service bus sdk listening for messages in a loop while True as per the docs for the sdk.I’m pretty sure a mre would be just the docs code running in a loop.As described in the issue we found the code change that introduced this behaviour.It might be desirable in some cases, but there needs to be some way of handling the disconnect or prevent the disconnect.Sent from my iPhoneOn 7 Feb 2024, at 16.31, github-actions[bot] ***@***.***> wrote:﻿
Hi @icfly2. Thank you for opening this issue and giving us the opportunity to assist. To help our team better understand your issue and the details of your scenario please provide a response to the question asked above or the information requested above. This will help us more accurately address your issue.

**Comment by l0lawrence** [comment]
Hi @icfly2,

`idle_timeout` was changed to `max_wait_time` when ServiceBus [migrated from v0.50 to v7.x](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/migration_guide.md#migrating-code-from-queueclient-and-receiver-to-servicebusreceiver-for-receiving-messages). Service Bus `max_wait_time` is used to control the time to wait for messages - see [here](https://github.com/Azure/azure-sdk-for-python/blob/86bb3152d480b2e9f70e083c6a5886025391849c/sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_client.py#L359).



Apologies, I'm not sure I fully understand the problem you are running into - are you trying to set an `idle_timeout`?

Could you please share the specific code sample you are running to help us to debug further? 

- Are you using `receive_messages` or the Iterator to receive? 
- Is this receiving on a session-based queue?
- Where were you setting `idle_timeout` when running v7.9.0?
- To confirm the disconnect error you mention above -- that was not present running the same sample in v7.9.0?

**Comment by icfly2** [comment]
> Hi @icfly2,
> 
> `idle_timeout` was changed to `max_wait_time` when ServiceBus [migrated from v0.50 to v7.x](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/migration_guide.md#migrating-code-from-queueclient-and-receiver-to-servicebusreceiver-for-receiving-messages). Service Bus `max_wait_time` is used to control the time to wait for messages - see [here](https://github.com/Azure/azure-sdk-for-python/blob/86bb3152d480b2e9f70e083c6a5886025391849c/sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_client.py#L359).
> 
> Apologies, I'm not sure I fully understand the problem you are running into - are you trying to set an `idle_timeout`?
No not really, I'd like the idle_timeout to be infinite or for the SDK to close and reopen connections when idle without raising and logging and exception which then ends up as an alert. And suppressing errors from the modules logger is obviously not a solution, though it has previously been suggested somwhere on github.
> 
> Could you please share the specific code sample you are running to help us to debug further?
Our implimentation follows [the official example](
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/samples/sync_samples/receive_subscription.py) very closely, only difference is we fetch connection strings from a keyvault and have an Enum for the subscriptions. This code however has not changed and the error will occur in the docs example if you let it run long enough, as by looking at the code snippets described in the issue description you can see that it checks for idle time and kills the connection if it has been idle for too long, but that the argument for what idel time is too long cannot be set as it does not propagate to the relevant function.
> 
>     * Are you using `receive_messages` or the Iterator to re ceive?
The implimentation uses the iterator implimentation:
```python
for message in service_bus_connection.reciever:
    ....
```

where `service_bus_connection.reciever` is the object returned by `azure.servicebus.ServiceBusSender` function `get_subscription_reciever`
> 
>     * Is this receiving on a session-based queue?
No just a bog standard service bus in the premium tier.
> 
>     * Where were you setting `idle_timeout` when running v7.9.0?
No, the option was not set.
> 
>     * To confirm, the disconnect error you mention above - that was not present running the same sample in v7.9.0?
 Yes, code has not changed, only sdk was bumped.

**Comment by l0lawrence** [comment]
Hi @icfly2,

I have so far been unable to reproduce the above error on my end - I am running the following sample with the latest ServiceBus version, as well as version 7.9.0:

```
CONNECTION_STR = os.environ['SERVICEBUS_CONNECTION_STR']
TOPIC_NAME = os.environ["SERVICEBUS_TOPIC_NAME"]
SUBSCRIPTION_NAME = os.environ["SERVICEBUS_SUBSCRIPTION_NAME"]

servicebus_client = ServiceBusClient.from_connection_string(conn_str=CONNECTION_STR)
with servicebus_client:
    receiver = servicebus_client.get_subscription_receiver(
        topic_name=TOPIC_NAME,
        subscription_name=SUBSCRIPTION_NAME
    )
    with receiver:
        for msg in receiver:
            print(str(msg))
            receiver.complete_message(msg)

print("Receive is done.")

```

Is it possible for you to provide debug level logs of what you are seeing?

**Comment by icfly2** [comment]
I've now got a new Error message in the log, very likely related to the same cause:

`Unable to attach new link: "ValueError('Invalid link')"`

```azure.servicebus._pyamqp._connection
Connection closed with error: [b'amqp:connection:forced', b"The connection was inactive for more than the allowed 60000 milliseconds and is closed by container 'LinkTracker'. TrackingId:1ae4434b6bfa48fda6ec5911bb5a7140_G6, SystemTracker:gateway5, Timestamp:2024-02-15T16:21:39", None]````

With a different total allowed timeout value 2x the other one.

I've also done some more digging and it appears that there need to be some messages on a regular basis on this bus for the mre to work. Some subscriptions that do not have a message on them don't trigger this, while there are two for which we get this in different environments. The service bus is configured the same everywhere...  (With TF)

**Comment by swathipil** [comment]
Hi @icfly2 - Sorry for the delay! When you bumped the version from 7.9.0 to any version above, did your receiving flow change in any way? For example, are there more/longer periods when the subscription is empty and you're waiting to receive?

The reason I ask is because the `b'amqp:connection:forced', b"The connection was inactive for more than the allowed 300000/60000 milliseconds` is a service-side error that happens when no messages have been received for that long. The timeout value is determined by the service and not something that's configured client-side. Whether it's v7.9.0 or later versions of the SDK, the client would not be able to keep the connection open if the service returns that error. As you mentioned, you can send a dummy message at least once in that time period and receive it to keep the connection open.

[[Here](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-troubleshoot#connection-is-closed)] are docs that describe the different scenarios that lead to different timeout values. These scenarios align with what you're seeing. The debug logs should more clearly show the behavior. You can track whether the connection/link exist by adding this line to your logging: `logging.basicConfig(level=logging.DEBUG, format='amqpConnection: %(amqpConnection)s, amqpSession: %(amqpSession)s, amqpLink: %(amqpLink)s')`

As for the `Unable to attach new link: "ValueError('Invalid link')"` error log, we've since updated this to a warning log and the update will be included in the upcoming release.

Thanks!

**Comment by swathipil** [comment]
@icfly2 - We'll also be updating the `connection closed with error` log from error to warning. Thanks @RichardOberdieck for helping with that!

**Comment by swathipil** [comment]
Error to warning log update has been released in version 7.12.0. Thanks!
